Add find and debug function

juraph-dev opened the pull request:
G'day,

This PR adds a simple debug head to the hydra, allowing a user to easily run a ros2 node through emacs-gdb. Works for local, and remote work-spaces through Tramp, including through a remote docker instance. 

I would like to extend the hydra with more options, but I wanted to gauge your thoughts on this before getting too far into it.

juraph-dev committed:
Add find and debug function

mbeutelspacher commented on ros.el:
```diff
@@ -598,6 +598,21 @@
+  "Search for, then launch a ROS2 node in GDB mode."
+  (interactive)
+  (if (eq (ros-current-version) 1) (error "Debug does not currently support ros1")
+  (let* ((lib-path (ros-shell-command-to-string "env | grep LD_LIBRARY_PATH"))
```
maybe rather do ``echo $LD_LIBRARY_PATH``?

juraph-dev replied:
An echo would be more elegant, however the output gets cut by the sed in 'ros-shell-command-to-string (I don't understand why)  and I figured the env | grep was less intrusive than modifying the expression, or creating a duplicate shell command function.

mbeutelspacher commented:
Thanks for this proposal, this definetly sounds like a good idea, I will try if I can also find a way for an ROS1 implementation

juraph-dev commented:
> Thanks for this proposal, this definitely sounds like a good idea, I will try if I can also find a way for an ROS1 implementation

Awesome! Since you're interested, I'll extend this for ROS1. Off the top of my head I know you can search nodelets with:
`rosrun nodelet declared_nodelets`, I'll have a look at how to handle nodes and see what  I can come up with. I'll mark this as draft until I implement it.

juraph-dev committed:
Add ros-1 find-debug functionality

Will check current ros version, using the corresponding
ros-find-debug-executable() function. Works for both nodes and nodelets,
presenting both to the user without additional intervention.

juraph-dev commented:
Added the ros1 debug function. Works well on my machine with a test ros1 workspace ([https://github.com/KumarRobotics/msckf_vio/tree/master](msckf_vio)) with both nodes and nodelets. GDB finding a source file for a Nodelet requires a *.so load at runtime, so you'll need to throw a blind breakpoint on first run ("b onInit()" is a safe bet). 

I am very new to lisp, so let me know if there is anything that should be cleaned.

juraph-dev commented:
@mbeutelspacher Friendly ping.

juraph-dev committed:
Add extra compile flags

juraph-dev committed:
Merge pull request #1 from mbeutelspacher/master

Bring up to date with head repo

juraph-dev committed:
cd to ws for compilation

Prevent colcon from surgically inserting build/install folders throughout my workspace. Instead, simply cd to workspace for compilation.

juraph-dev committed:
Change finder for ros2 executablers

I ran into a few issues with certain packages, notably sub-packages not being identified by ros-list-packages. Fix is to use ros2 list executables, trimming the output and using that as the pipe to gdb.

juraph-dev committed:
Merge remote-tracking branch 'upstream/master'

juraph-dev committed:
Merge remote-tracking branch 'origin/devel'

juraph-dev committed:
Add "last debug action"

Allow the user to repeat the last gdb command, instead of having to
re-select the package, etc. Bound to _l_.

juraph-dev committed:
Update Readme

Attempting to justify the fork.

juraph-dev committed:
Change method for filtering executables

Sometimes, people (me) will create an executable with the same name as
its package. Handle appropriately.

juraph-dev committed:
Merge pull request #4 from mbeutelspacher/master

stealing latest updates.

juraph-dev committed:
Drop support for 27.1

Upgrades people, upgrades!

juraph-dev committed:
Merge remote-tracking branch 'upstream/master'

Cheers Max!
